rdock-utils/rdock_utils/rbhtfinder/rbhtfinder.py

```diff
@@ -0,0 +1,357 @@
+            column_names = [f"COL{n+1}" for n in range(len(sdreport_dataframe.columns))]
+
+        sdreport_array = sdreport_dataframe.values
+        return sdreport_array, column_names
+
+    def read_data_using_numpy(self) -> InputData:
+        np_array = np.loadtxt(self.config.input, dtype=str)
+
+        if self.config.header:
+            column_names = np_array[0]
+            sdreport_array = np_array[1:]
+        else:
+            column_names = [f"COL{n+1}" for n in range(np_array.shape[1])]
+            sdreport_array = np_array
+
+        return sdreport_array, column_names
+
+    def prepare_array(self, data_array: SDReportArray, name_column: int) -> Array3DFloat:
+        """
+        Convert `sdreport_array` (read directly from the tsv) to 3D array (molecules x poses x columns) and filter out molecules with too few/many poses
+        """
+        split_indices = (
+            np.where(
+                data_array[:, name_column] != np.hstack((data_array[1:, name_column], data_array[0, name_column]))
+            )[0]
+            + 1
+        )
+        split_array = np.split(data_array, split_indices)
+        modal_shape = Counter([array.shape for array in split_array]).most_common(1)[0]
+        number_of_poses = modal_shape[0][0]  # Find modal number of poses per molecule in the array
+        valid_split_arrays = [
+            np.array_split(array, array.shape[0] / number_of_poses)  # type: ignore
+            for array in split_array
+            if not array.shape[0] % number_of_poses and array.shape[0]
+        ]
+        flattened_split_array = np.concatenate(valid_split_arrays)
+
+        if len(flattened_split_array) * number_of_poses < data_array.shape[0] * 0.99:
+            message = (
+                "WARNING: The number of poses provided per molecule is inconsistent. "
+                f"Only {len(flattened_split_array)} of {int(data_array.shape[0] / number_of_poses)} molecules have {number_of_poses} poses."
+            )
+            logger.warning(message)
+
+        molecule_3d_array = np.array(flattened_split_array)
+        # Overwrite the name column (should be the only one with dtype=str) so we can force everything to float
+        molecule_3d_array[:, :, name_column] = 0
+        final_array = molecule_3d_array.astype(float)
+        return final_array
+
+    def process_filter_combinations(
+        self, molecule_array: Array3DFloat, distinct_combinations: Array2DFloat
+    ) -> list[FilterCombinationResult]:
+        # Find the top scoring compounds for validation of the filter combinations
+        columns = set(filter.column for filter in self.config.filters)
+        min_score_indices = {
+            column: np.argpartition(np.min(molecule_array[:, :, column], axis=1), self.config.validation)[
+                : self.config.validation
+            ]
+            for column in columns
+        }
+        num_cpus = os.cpu_count() or 1
```

Comment:
the number of cpus should be part of the config structure, and be passed via command line argument.  

let's say something like `-c` / `--cpu-count` with a default of 1.